### PR TITLE
Add median filtering and time trim

### LINKED ISSUE #75: Add Test Length Slider to filter data by time
To reduce regression dilution.

### PR DESCRIPTION
**Pull request by Piphi5**
Applies a median filter on the velocity data to remove outliers
Incorporates time trimming to allow teams to better trim their dynamic test data.
Resolves #75 

Issues:
- [x] Fix Graph scaling
- [x] Refactor Analyzer a bit more
- [x] Add tests?
- [x] Add graph legend
- [x] Setup Filtered, Raw, Simulate Graphs stuff

This should allow teams to significantly clean up noisy data.
Here's an example using the integration test flywheel sim with encoder noise set to 0.1:

### Time domain graphs (Left is current tool, Right is PR):
image

### Voltage Domain Graphs (Left is current tool, Right is PR):
image

### Gains and Settings (Top is current tool, Bottom is PR):
image

### CODE REVIEW

**Comment by Piphi5** [comment]
Updated Graphs Setup:
image
image
image
image

**Comment by calcmogul** [comment]
Make sure raw data is displayed before the filtered data so it gets put behind the filtered data. Some of the plots do the opposite.

**Comment by Piphi5** [comment]
Does this setup work?
image

File: src/main/native/cpp/view/AnalyzerPlot.cpp [hunk 1]
```diff
@@ -248,31 +275,73 @@ void AnalyzerPlot::DisplayTimeDomainPlots() {
...
+      // Plot simulated time-domain data or raw data.
+      if (i == 2 || (i == 4)) {
+        // Plot Filtered Data before sim data
+        ImPlot::SetNextMarkerStyle(IMPLOT_AUTO, 1, IMPLOT_AUTO_COL, 0);
+        ImPlot::PlotScatterG("Filtered Data", Getter, filteredData.data(),
+                             filteredData.size());
+        if (i == 2) {
+          for (auto&& pts : m_quasistaticSim) {
+            ImPlot::SetNextLineStyle(IMPLOT_AUTO_COL, 1.5);
+            ImPlot::PlotLineG("Simulated", Getter, pts.data(), pts.size());
```
Comment (calcmogul) [inline src/main/native/cpp/view/AnalyzerPlot.cpp:301]: Simulated what? Maybe just say "Simulation".

File: src/main/native/include/sysid/analysis/FilteringUtils.h [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+/**
+ * Trims the step voltage data to discard all points before the maximum
+ * acceleration.
+ *
+ * @param data A pointer to the step voltage data.
+ * @param stepTestDuration A reference to the step test duration that will be
+ * used.
+ * @param minStepTime A reference to the minimum step time that will be used.
+ */
+inline void TrimStepVoltageData(std::vector<PreparedData>* data,
```
Comment (calcmogul) [inline src/main/native/include/sysid/analysis/FilteringUtils.h:82]: Rather than using `inline`, we should move the implementation of this function to a FilteringUtils.cpp.

File: src/main/native/cpp/view/AnalyzerPlot.cpp [hunk 3]
```diff
@@ -147,11 +155,30 @@ void AnalyzerPlot::SetData(const Storage& data, const std::vector<double>& ff,
...
+  // Populate Raw Slow Time Series Data
+  for (size_t i = 0; i < raw_slow.size(); i += raw_sStep) {
+    m_rawData[kChartTitles[2]].emplace_back(raw_slow[i].timestamp - t,
+                                            raw_slow[i].velocity);
+    m_rawData[kChartTitles[3]].emplace_back(raw_slow[i].timestamp - t,
+                                            raw_slow[i].acceleration);
+  }
+
+  t = raw_fast[0].timestamp;
+  // Populate Raw fast Time Series Data
```
Comment (calcmogul) [inline src/main/native/cpp/view/AnalyzerPlot.cpp:176]: The comment casing here is inconsistent.

File: src/main/native/cpp/view/AnalyzerPlot.cpp [hunk 4]
```diff
@@ -248,31 +275,73 @@ void AnalyzerPlot::DisplayTimeDomainPlots() {
...
     if (ImPlot::BeginPlot(kChartTitles[i], x, y, ImVec2(-1, 0),
                           ImPlotFlags_None, ImPlotAxisFlags_NoGridLines)) {
-      // Get a reference to the data we are plotting.
-      auto& data = m_data[kChartTitles[i]];
+      // Set Legend Location:
+      ImPlot::SetLegendLocation(ImPlotLocation_East, ImPlotOrientation_Vertical,
+                                true);
+
+      // Plot simulated time-domain data or raw data.
+      if (i == 2 || (i == 4)) {
```
Comment (calcmogul) [inline src/main/native/cpp/view/AnalyzerPlot.cpp:293]: This whole if structure could use refactoring for readability and reduced duplication.

File: src/main/native/include/sysid/analysis/AnalysisManager.h [hunk 5]
```diff
@@ -159,26 +161,39 @@ class AnalysisManager {
    *
    * @return A reference to the raw internal data.
    */
-  Storage& GetRawData() { return m_datasets[kDatasets[m_settings.dataset]]; }
+  Storage& GetRawData() { return m_rawDatasets[kDatasets[m_settings.dataset]]; }
```
Comment (calcmogul) [inline src/main/native/include/sysid/analysis/AnalysisManager.h:164]: Put a newline after this.

File: src/main/native/include/sysid/analysis/AnalysisManager.h [hunk 6]
```diff
@@ -159,26 +161,39 @@ class AnalysisManager {
...
+  Storage& GetFilteredData() {
+    return m_filteredDatasets[kDatasets[m_settings.dataset]];
+  }
+
+  double GetMinDuration() const { return m_minDuration; }
+
+  double GetMaxDuration() const { return m_maxDuration; }
+
+  void UpdateSliderValues();
```
Comment (calcmogul) [inline src/main/native/include/sysid/analysis/AnalysisManager.h:173]: Write documentation for these functions.

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Comment by calcmogul** [comment]
The median filter is cropping a lot of data and causing strange results.
Screenshot_2021-03-24_16-07-07

File: src/main/native/include/sysid/analysis/FilteringUtils.h [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+ * Reduces noise in velocity data by applying a median filter.
+ *
+ * @tparam S The size of the raw data array
+ * @tparam Velocity The index of the velocity entry in the raw data.
+ *
+ * @param data the prepared data vector containing acceleration data
+ * @param window the size of the window of the median filter (must be odd)
+ */
+template <size_t S, size_t Velocity>
+std::vector<std::array<double, S>> ApplyMedianFilter(
```
Comment (calcmogul) [inline src/main/native/include/sysid/analysis/FilteringUtils.h:47]: Can we use MedianFilter in WPILib for this?

**Comment by Piphi5** [comment]
> We may have to change how we handle the boundary conditions, like doing a median filter for the remaining data but with fewer points rather than dropping the boundary data.

Could we do zero padding?

**Comment by calcmogul** [comment]
If we were filtering acceleration, zero padding would work. It would do weird things to velocity data though.

File: src/main/native/include/sysid/analysis/FilteringUtils.h [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+  size_t step = window / 2;
+
+  std::vector<std::array<double, S>> prepared;
+
+  // Run median filter on non-edges
+  for (int i = step; i < data.size() - step; i++) {
+    std::vector<std::array<double, S>> velocity_window(window);
+    velocity_window.assign(data.begin() + i - step,
+                           data.begin() + i + step + 1);
+    std::sort(velocity_window.begin(), velocity_window.end(),
```
Comment (calcmogul) [inline src/main/native/include/sysid/analysis/FilteringUtils.h:62]: Use `thisCase` for variables, not `this_case`. There's other places in this PR too.

**Comment by Piphi5** [comment]
> If we were filtering acceleration, zero padding would work. It would do weird things to velocity data though.

would grabbing whatever the median filter calculates for those values work?

File: src/test/native/cpp/analysis/FilterTest.cpp [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+// the WPILib BSD license file in the root directory of this project.
+
+#include <array>
+#include <vector>
+
+#include "gtest/gtest.h"
+#include "sysid/analysis/FeedforwardAnalysis.h"
+#include "sysid/analysis/FilteringUtils.h"
+
+TEST(FilterTests, MedianFilter) {
```
Comment (calcmogul) [inline src/test/native/cpp/analysis/FilterTest.cpp:12]:
```suggestion
TEST(FilterTest, MedianFilter) {
```

File: src/test/native/cpp/analysis/FilterTest.cpp [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  std::vector<std::array<double, 1>> test_data = {{0}, {1},    {10}, {5}, {3},
+                                                  {0}, {1000}, {7},  {6}, {5}};
+  std::vector<std::array<double, 1>> expected_data = {{1}, {5}, {5}, {3},
+                                                      {3}, {7}, {7}, {6}};
+  std::vector<std::array<double, 1>> filtered_data =
+      sysid::ApplyMedianFilter<1, 0>(test_data, 3);
+  EXPECT_EQ(expected_data, filtered_data);
+}
+
+TEST(FilterTests, QuasistaticTrim) {
```
Comment (calcmogul) [inline src/test/native/cpp/analysis/FilterTest.cpp:22]:
```suggestion
TEST(FilterTest, QuasistaticTrim) {
```

File: src/test/native/cpp/analysis/FilterTest.cpp [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  std::vector<std::array<double, 2>> test_data = {
+      {0, 1}, {.5, 2}, {2, 0.1}, {4, 4}, {0, 5}};
+  std::vector<std::array<double, 2>> expected_data = {{.5, 2}, {4, 4}};
+  sysid::TrimQuasistaticData<2, 0, 1>(&test_data, 0.2);
+  EXPECT_EQ(expected_data, test_data);
+}
+
+TEST(FilterTests, StepTrim) {
```
Comment (calcmogul) [inline src/test/native/cpp/analysis/FilterTest.cpp:30]:
```suggestion
TEST(FilterTest, StepTrim) {
```

**Review by calcmogul — CHANGES_REQUESTED** [review]
Not all the plots are zoomable, by the way.

**Comment by calcmogul** [comment]
> would grabbing whatever the median filter calculates for those values work?

Well, the issue is there's no sane data to use for the boundary condition. The default median filter impl isn't really defined there.

File: src/test/native/cpp/analysis/FilterTest.cpp [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+#include <wpi/raw_ostream.h>
+
+#include "gtest/gtest.h"
+#include "sysid/analysis/AnalysisManager.h"
+#include "sysid/analysis/FeedforwardAnalysis.h"
+#include "sysid/analysis/FilteringUtils.h"
+
+TEST(FilterTest, MedianFilter) {
+  std::vector<std::array<double, 1>> test_data = {{0}, {1},    {10}, {5}, {3},
```
Comment (calcmogul) [inline src/test/native/cpp/analysis/FilterTest.cpp:16]: Use mixed case instead of snake case for variables.

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: src/main/native/include/sysid/view/Analyzer.h [hunk 13]
```diff
@@ -86,5 +86,9 @@ class Analyzer : public glass::View {
 
   // Plot
   AnalyzerPlot m_plot{m_logger};
+
+  float m_graphScroll = 0.0;
+  bool m_graphLoaded = false;
+  bool m_dataLoaded = false;
```
Comment (calcmogul) [inline src/main/native/include/sysid/view/Analyzer.h:92]: Since this is modified in the thread.
```suggestion
  std::atomic<bool> m_dataLoaded{false};
```

File: src/main/native/cpp/view/Analyzer.cpp [hunk 14]
```diff
@@ -231,31 +199,49 @@ void Analyzer::Display() {
...
+      m_plot.DisplayTimeDomainPlots();
 
-      // Show time domain diagnostic plots.
-      if (ImGui::BeginPopupModal("Time-Domain Diagnostics")) {
-        m_plot.DisplayTimeDomainPlots();
-        // Button to close popup.
-        if (ImGui::Button("Close")) {
-          ImGui::CloseCurrentPopup();
+      if (m_dataLoaded) {
+        if (m_graphLoaded) {
```
Comment (calcmogul) [inline src/main/native/cpp/view/Analyzer.cpp:237]: What's the difference between m_dataLoaded and m_graphLoaded? Seems like m_graphLoaded just follows whatever m_dataLoaded does.

The purpose of this logic is to be an edge detector: "if the data wasn't loaded before and it is now".

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: src/main/native/cpp/view/Analyzer.cpp (hunk 14 above)
Comment (Piphi5) [reply src/main/native/cpp/view/Analyzer.cpp:237]: should it be prevDataLoaded?

Comment (calcmogul) [reply src/main/native/cpp/view/Analyzer.cpp:237]: I think so. It makes it clearer what it's for.

File: src/main/native/include/sysid/analysis/FilteringUtils.h [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+// the WPILib BSD license file in the root directory of this project.
+
+#pragma once
+
+#include <algorithm>
+#include <array>
+#include <cmath>
+#include <vector>
+
+#include <frc/LinearFilter.h>
```
Comment (calcmogul) [inline src/main/native/include/sysid/analysis/FilteringUtils.h:12]: `#include <frc/LinearFilter.h>` can go in FilteringUtils.cpp.

**Review by calcmogul — CHANGES_REQUESTED** [review]

**Comment by calcmogul** [comment]
This needs a formatter run on it.

File: src/test/native/cpp/analysis/FilterTest.cpp [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+      {units::second_t{0}, 1, 2, 3, 0, 0},
+      {units::second_t{1}, 1, 2, 3, 1, 0},
+      {units::second_t{2}, 1, 2, 3, 2, 0},
+      {units::second_t{3}, 1, 2, 3, 5, 0},
+      {units::second_t{4}, 1, 2, 3, 0.35, 0},
+      {units::second_t{5}, 1, 2, 3, 0.15, 0},
+      {units::second_t{6}, 1, 2, 3, 0, 0},
+      {units::second_t{7}, 1, 2, 3, 0.02, 0},
+      {units::second_t{8}, 1, 2, 3, 0.01, 0},
+      {units::second_t{9}, 1, 2, 3, 0, 0}};
```
Comment (calcmogul) [inline src/test/native/cpp/analysis/FilterTest.cpp:44]:
```suggestion
      {0_s, 1, 2, 3, 0, 0},
      {1_s, 1, 2, 3, 1, 0},
      {2_s, 1, 2, 3, 2, 0},
      {3_s, 1, 2, 3, 5, 0},
      {4_s, 1, 2, 3, 0.35, 0},
      {5_s, 1, 2, 3, 0.15, 0},
      {6_s, 1, 2, 3, 0, 0},
      {7_s, 1, 2, 3, 0.02, 0},
      {8_s, 1, 2, 3, 0.01, 0},
      {9_s, 1, 2, 3, 0, 0}};
```

File: src/test/native/cpp/analysis/FilterTest.cpp [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+      {units::second_t{0}, 1, 2, 3, 0, 0},
+      {units::second_t{1}, 1, 2, 3, 0.25, 0},
+      {units::second_t{2}, 1, 2, 3, 0.5, 0},
+      {units::second_t{3}, 1, 2, 3, 0.45, 0},
+      {units::second_t{4}, 1, 2, 3, 0.35, 0},
+      {units::second_t{5}, 1, 2, 3, 0.15, 0},
+      {units::second_t{6}, 1, 2, 3, 0, 0},
+      {units::second_t{7}, 1, 2, 3, 0.02, 0},
+      {units::second_t{8}, 1, 2, 3, 0.01, 0},
+      {units::second_t{9}, 1, 2, 3, 0, 0},
```
Comment (calcmogul) [inline src/test/native/cpp/analysis/FilterTest.cpp:60]:
```suggestion
      {0_s, 1, 2, 3, 0, 0},
      {1_s, 1, 2, 3, 0.25, 0},
      {2_s, 1, 2, 3, 0.5, 0},
      {3_s, 1, 2, 3, 0.45, 0},
      {4_s, 1, 2, 3, 0.35, 0},
      {5_s, 1, 2, 3, 0.15, 0},
      {6_s, 1, 2, 3, 0, 0},
      {7_s, 1, 2, 3, 0.02, 0},
      {8_s, 1, 2, 3, 0.01, 0},
      {9_s, 1, 2, 3, 0, 0},
```

File: src/test/native/cpp/analysis/FilterTest.cpp [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+  };
+
+  std::vector<sysid::PreparedData> expectedData = {
+      {units::second_t{2}, 1, 2, 3, 0.5, 0},
+      {units::second_t{3}, 1, 2, 3, 0.45, 0},
+      {units::second_t{4}, 1, 2, 3, 0.35, 0},
+      {units::second_t{5}, 1, 2, 3, 0.15, 0}};
```
Comment (calcmogul) [inline src/test/native/cpp/analysis/FilterTest.cpp:67]:
```suggestion
      {2_s, 1, 2, 3, 0.5, 0},
      {3_s, 1, 2, 3, 0.45, 0},
      {4_s, 1, 2, 3, 0.35, 0},
      {5_s, 1, 2, 3, 0.15, 0}};
```

File: src/test/native/cpp/analysis/FilterTest.cpp [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+
+  auto maxTime = units::second_t{9};
```
Comment (calcmogul) [inline src/test/native/cpp/analysis/FilterTest.cpp:69]:
```suggestion
  auto maxTime = 9_s;
```

**Review by calcmogul — CHANGES_REQUESTED** [review]

File: src/main/native/cpp/analysis/FilteringUtils.cpp [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+    auto endIt = std::find_if(
+        std::reverse_iterator{data->end()},
+        std::reverse_iterator{data->begin()}, [&](const PreparedData& entry) {
+          return std::abs(entry.acceleration) > accelNoiseFloor;
+        });
+
+    // Calculate default duration
+    settings.stepTestDuration =
+        std::min(endIt->timestamp - data->front().timestamp + minStepTime +
+                     units::second_t{1},
```
Comment (calcmogul) [inline src/main/native/cpp/analysis/FilteringUtils.cpp:42]:
```suggestion
                     1_s,
```

**Review by calcmogul — APPROVED** [review]
